test/controllers/UpdateRelationshipControllerTest.scala

```diff
@@ -18,500 +18,780 @@ package controllers
+
+      when(mockUpdateRelationshipService.getMAEndingDatesForDivorce(role, divorceDate))
+        .thenReturn(Future.successful(maEndingDates))
+
+      when(mockUpdateRelationshipService.saveMarriageAllowanceEndingDates(ArgumentMatchers.eq(maEndingDates))(any()))
+        .thenReturn(Future.successful(maEndingDates))
+
+      val viewModel = DivorceEndExplanationViewModel(role, divorceDate, maEndingDates)
+
+      val result = controller().divorceEndExplanation()(request)
+      status(result) shouldBe OK
+
+      result rendersTheSameViewAs views.html.coc.divorce_end_explanation(viewModel)
+
+    }
+
+    "display an error page" when {
+
+      "an error occurs retrieving divorce data" in {
+
+        when(mockUpdateRelationshipService.getDataForDivorceExplanation(any(), any()))
+          .thenReturn(Future.failed(CacheMissingRelationshipRecords()))
+
+        val result = controller().divorceEndExplanation()(request)
+        status(result) shouldBe INTERNAL_SERVER_ERROR
+
+        result rendersTheSameViewAs views.html.errors.try_later()
+      }
+
+      "an error has occurred whilst saving cache data" in {
+
+        val role = Transferor
+        val divorceDate = LocalDate.now().minusDays(1)
+        val maEndingDate = LocalDate.now().plusDays(1)
+        val paEffectiveDate = LocalDate.now().plusDays(2)
+
+        val maEndingDates = MarriageAllowanceEndingDates(maEndingDate, paEffectiveDate)
+
+        when(mockUpdateRelationshipService.getDataForDivorceExplanation(any(), any()))
+          .thenReturn(Future.successful((role, divorceDate)))
+
+        when(mockUpdateRelationshipService.getMAEndingDatesForDivorce(role, divorceDate))
+          .thenReturn(Future.successful(maEndingDates))
+
+        when(mockUpdateRelationshipService.saveMarriageAllowanceEndingDates(ArgumentMatchers.eq(maEndingDates))(any()))
+          .thenReturn(failedFuture)
+
+        val result = controller().divorceEndExplanation()(request)
+        status(result) shouldBe INTERNAL_SERVER_ERROR
+
+        result rendersTheSameViewAs views.html.errors.try_later()
       }
     }
+
   }
 
   "confirmEmail" should {
-    "return a success" when {
+    "display the confirm email page" when {
       "an email is recovered from the cache" in {
         val email = "test@test.com"
-        when(mockUpdateRelationshipService.getUpdateNotification(any(), any()))
-          .thenReturn(Some(NotificationRecord(EmailAddress(email))))
-        val result = controller().confirmEmail()(request)
+
+        when(mockUpdateRelationshipService.getEmailAddress(any(), any()))
+          .thenReturn(Future.successful(Some(email)))
+
+        val result = controller().confirmEmail(request)
         status(result) shouldBe OK
-        val document = Jsoup.parse(contentAsString(result))
-        document.getElementById("transferor-email").attr("value") shouldBe email
+
+        result rendersTheSameViewAs views.html.coc.email(emailForm.fill(EmailAddress(email)))
```

Comment:
this feels a bit 'intergrationy'

Reply:
Not sure what you mean by this. Are you referring to rendersTheSameViewAs or the way the email view is populated?

Reply:
the test feels like its testing more than just the controller, but also the view

Reply:
There is an element of needing to provide the correct parameters to the view whether that be a viewmodel or individual arguments. After that is testing the fact the controller is routing us to the correct view, something the other controller tests are not doing. There are no elements inside the view that are being evaluated.

Reply:
as disscussed